docs/idiomatic-maya-guide.md

```diff
@@ -0,0 +1,418 @@
+
+## If Maya is Go then why not Idiomatic Go
+Being idiomatic in Maya includes all the idiomatic pieces in Go and also takes into consideration writing code that understands and responds to Kubernetes a lot better. In addition, team at Maya has tried to put all their storage learnings so far into what is being termed as Idiomatic Maya where Maya is control plane for OpenEBS.
+
+
+## Devil is in the details
+Below are some sample code reviews that describes with examples to write idiomatic maya code.
+
+### Naming - Thumb Rule
+- Use names based on what the logic provides and not based on what the logic contains
+
+```go
+import (
+  // alias `algorithm` conveys about what it contains i.e. some algorithm
+  algorithm "github.com/openebs/maya/pkg/algorithm/nodeSelect/v1alpha1"
+
+  // vs.
+
+  // `nodeselect` tries to convey what the logic provides
+  // this seems more natural way to express
+  nodeselect "github.com/openebs/maya/pkg/algorithm/nodeSelect/v1alpha1"
+)
+```
+
+```go
+nodeDisks, err := ac.NodeDiskSelector()
+// `disks` is repeated; does not seem natural
+if len(nodeDisks.Disks.Items) == 0 {
+  // ...
+}
+
+// vs.
+
+s, err := ac.NodeDiskSelector()
+if len(s.Disks.Items) == 0 {
+  // ...
+}
+```
+
+### Do the names reflect their purpose?
+```go
+// assume a file pkg/algorithm/nodeSelect/v1alpha1/select_node.go 
+// contains below code
+
+// NodeDiskSelector selects a node and disks attached to it.
+//
+// Method name clashes with package name
+// These are few questions that comes to mind:
+//  1/ Does this logic return a node instance?
+//  2/ Does this logic return a disk instance?
+//
+// However, comment does help to some extent in understanding the purpose 
+// of this logic
+func (ac *AlgorithmConfig) NodeDiskSelector() (*nodeDisk, error) {
+	listDisk, err := ac.getDisk()
+	if listDisk == nil || len(listDisk.Items) == 0 {
+		return nil, errors.Wrapf(err, "no disk object found")
+	}
+	nodeDiskMap, err := ac.getCandidateNode(listDisk)
+	if err != nil {
+		return nil, err
+	}
+	selectedDisk := ac.selectNode(nodeDiskMap)
+	return selectedDisk, nil
+}
+
+vs.
+
+// change the file path to below
+// pkg/nodeselect/v1alpha1/nodeselect.go
+
+// NodeDiskSelector selects and returns appropriate node
+//
+// Note: It is assumed that the information required to select is 
+// available in nodeselect instance
+//
+// Note: The return has to be an instance of node. This instance
+// should be defined in this package
+//
+// Note: Below might indicate a stripped off version, however it is
+// not. Code needs to be designed to place business logic in proper places.
+// A good code reflects in its Unit Testing as well as in its caller code.
+func (s *nodeselect) Get() (*node, error) {
+	d, err := s.getDisks()
+	if d == nil || len(d.Items) == 0 {
+		return nil, errors.Wrapf(err, "no disks provided")
+	}
+	dl := disk.NewList(disk.WithNames(d.Items)).Filter(disk.IsFree)
+	nl := node.NewList(node.WithDisks(dl.Item)).Filter(node.ContainsDisk)
```

Comment:
I find it hard to digest above line, there is no error handling in above code also?
IMO, putting an example of `WithDisks`, `Filter` above to this example will help.